homeassistant/components/withings/__init__.py

```diff
@@ -59,52 +82,124 @@ async def async_setup(hass: HomeAssistantType, config: ConfigType) -> bool:
     return True
 
 
-async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry) -> bool:
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up Withings from a config entry."""
-    # Upgrading existing token information to hass managed tokens.
-    if "auth_implementation" not in entry.data:
-        _LOGGER.debug("Upgrading existing config entry")
-        data = entry.data
-        creds = data.get(const.CREDENTIALS, {})
-        hass.config_entries.async_update_entry(
-            entry,
-            data={
-                "auth_implementation": const.DOMAIN,
-                "implementation": const.DOMAIN,
-                "profile": data.get("profile"),
-                "token": {
-                    "access_token": creds.get("access_token"),
-                    "refresh_token": creds.get("refresh_token"),
-                    "expires_at": int(creds.get("token_expiry")),
-                    "type": creds.get("token_type"),
-                    "userid": creds.get("userid") or creds.get("user_id"),
-                },
+    # Normalize the config entry.
+    webhook_id = entry.data.get(CONF_WEBHOOK_ID, webhook.async_generate_id())
+    hass.config_entries.async_update_entry(
+        entry,
+        unique_id=entry.data["token"]["userid"],  # Upgrade existing configs.
+        data={
+            **entry.data,
+            **{
+                const.CLIENT_ID: hass.data[DOMAIN][const.CONFIG][const.CLIENT_ID],
+                const.CLIENT_SECRET: hass.data[DOMAIN][const.CONFIG][
+                    const.CLIENT_SECRET
+                ],
+                const.USE_WEBHOOK: hass.data[DOMAIN][const.CONFIG][const.USE_WEBHOOK],
+                CONF_WEBHOOK_ID: webhook_id,
+                const.CONF_WEBHOOK_URL: entry.data.get(
+                    const.CONF_WEBHOOK_URL, webhook.async_generate_url(hass, webhook_id)
+                ),
```

Comment:
we shouldn't update on each setup. We only should update if we have new info that is not stored yet in the config entry.